apiserver/internal/handlers/resource/resourcemigration.go

```diff
@@ -139,27 +125,42 @@ func (h *resourcesMigrationUploadHandler) processPost(
 			return empty, internalerrors.Capture(err)
 		}
 	}
+
 	if !isPlaceholder(query) {
 		var (
 			retrievedBy     string
-			retrievedByType coreresource.RetrievedByType
+			retrievedByType resource.RetrievedByType
 		)
 		if target.unitUUID != "" {
 			retrievedBy = target.unitUUID.String()
-			retrievedByType = coreresource.Unit
+			retrievedByType = resource.Unit
 		} else if userID != "" {
 			retrievedBy = userID
-			retrievedByType = coreresource.User
+			retrievedByType = resource.User
 		} else {
 			retrievedBy = target.appID.String()
-			retrievedByType = coreresource.Application
+			retrievedByType = resource.Application
+		}
+
+		details, err := resourceDetailsFromQuery(query)
+		if err != nil {
+			return empty, internalerrors.Errorf("extracting resource details from request: %w", err)
+		}
+
+		reader, err := h.validator.Validate(r.Body, details.fingerprint.String(), details.size)
```

Comment:
How can a validator return a reader? Validation returns an error, nothing more.

Reply:
Changed name to ValidateAndStoreReader.